File: chronik/bitcoinsuite-core/src/tx/tx.rs
// Copyright (c) 2023 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

use crate::{
    error::DataError,
    script::Script,
    ser::{BitcoinSer, BitcoinSerializer},
    tx::TxId,
};

/// CTransaction, a Bitcoin transaction.
///
/// ```
/// # use bitcoinsuite_core::tx::{Tx, TxId, TxMut};
/// let txid = TxId::from([3; 32]);
/// let tx = Tx::with_txid(
///     txid,
///     TxMut {
///         version: 1,
///         inputs: vec![],
///         outputs: vec![],
///         locktime: 0,
///     },
/// );
/// assert_eq!(tx.txid(), txid);
/// assert_eq!(tx.version, 1);
/// assert_eq!(tx.inputs, vec![]);
/// assert_eq!(tx.outputs, vec![]);
/// assert_eq!(tx.locktime, 0);
/// ```
///
/// Immutable version of [`TxMut`], so this will fail:
/// ```compile_fail
/// # use bitcoinsuite_core::tx::Tx;
/// let mut tx = Tx::default();
/// tx.version = 1;
/// ```
#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Tx {
    txid: TxId,
    tx: TxMut,
}

/// Bitcoin transaction. Mutable version of [`Tx`], like CMutableTransaction.
///
/// The biggest difference is that it doesn't have a txid() method, which we
/// cannot know without hashing the tx every time, which would be expensive to
/// compute.
#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct TxMut {
    /// nVersion of the tx.
    pub version: i32,
    /// Tx inputs.
    pub inputs: Vec<TxInput>,
    /// Tx outputs.
    pub outputs: Vec<TxOutput>,
    /// Locktime of the tx.
    pub locktime: u32,
}

/// COutPoint, pointing to a coin being spent.
#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct OutPoint {
    /// TxId of the output of the coin.
    pub txid: TxId,
    /// Index in the outputs of the tx of the coin.
    pub out_idx: u32,
}

/// Points to an input spending a coin.
#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct SpentBy {
    /// TxId of the tx with the input.
    pub txid: TxId,
    /// Index in the inputs of the tx.
    pub input_idx: u32,
}

/// CTxIn, spending an unspent output.
#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct TxInput {
    /// Points to an output being spent.
    pub prev_out: OutPoint,
    /// scriptSig unlocking the output.
    pub script: Script,
    /// nSequence.
    pub sequence: u32,
    /// Coin being spent by this tx.
    /// May be [`None`] for coinbase txs or if the spent coin is unknown.
    pub coin: Option<Coin>,
}

/// CTxOut, creating a new output.
#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct TxOutput {
    /// Value of the output.
    pub value: i64,
    /// Script locking the output.
    pub script: Script,
}

/// Coin, can be spent by providing a valid unlocking script.
#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Coin {
    /// Output, locking the coins.
    pub output: TxOutput,
    /// Height of the coin in the chain.
    pub height: i32,
    /// Whether the coin is a coinbase.
    pub is_coinbase: bool,
}

/// Empty tx, serializes to 00000000000000000000.
pub static EMPTY_TX: TxMut = TxMut {
    version: 0,
    inputs: vec![],
    outputs: vec![],
    locktime: 0,
};

impl OutPoint {
    /// Whether this outpoint is a coinbase input.
    ///
    /// ```
    /// # use bitcoinsuite_core::tx::{OutPoint, TxId};
    /// let txid0 = TxId::from([0; 32]);
    /// let txid3 = TxId::from([3; 32]);
    ///
    /// fn is_coinbase(txid: TxId, out_idx: u32) -> bool {
    ///     OutPoint { txid, out_idx }.is_coinbase()
    /// }
    ///
    /// assert!(!is_coinbase(txid3, u32::MAX));
    /// assert!(!is_coinbase(txid3, 0));
    /// assert!(!is_coinbase(txid0, 0));
    /// assert!(!is_coinbase(txid0, u32::MAX - 1));
    /// assert!(is_coinbase(txid0, u32::MAX));
    /// ```
    pub fn is_coinbase(&self) -> bool {
        self.txid == TxId::default() && self.out_idx == u32::MAX
    }
}

impl Tx {
    /// Create a new [`Tx`] with a given [`TxId`] and [`TxMut`].
    ///
    /// It is the responsibility of the caller to ensure the `txid` matches
    /// `tx`.
    pub fn with_txid(txid: TxId, tx: TxMut) -> Self {
        Tx { txid, tx }
    }

    /// [`TxId`] of this [`Tx`].
    pub fn txid(&self) -> TxId {
        self.txid
    }

    /// Like [`Tx::txid`], but as a reference.
    pub fn txid_ref(&self) -> &TxId {
        &self.txid
    }
}

impl std::ops::Deref for Tx {
    type Target = TxMut;

    fn deref(&self) -> &Self::Target {
        &self.tx
    }
}

impl BitcoinSer for TxMut {
    fn ser_to<S: BitcoinSerializer>(&self, bytes: &mut S) {
        self.version.ser_to(bytes);
        self.inputs.ser_to(bytes);
        self.outputs.ser_to(bytes);
        self.locktime.ser_to(bytes);
    }

    fn deser(data: &mut bytes::Bytes) -> Result<Self, DataError> {
        Ok(TxMut {
            version: BitcoinSer::deser(data)?,
            inputs: BitcoinSer::deser(data)?,
            outputs: BitcoinSer::deser(data)?,
            locktime: BitcoinSer::deser(data)?,
        })
    }
}

impl BitcoinSer for Tx {
    fn ser_to<S: BitcoinSerializer>(&self, bytes: &mut S) {
        TxMut::ser_to(self, bytes)
    }

    fn deser(data: &mut bytes::Bytes) -> Result<Self, DataError> {
        let tx = TxMut::deser(data)?;
        Ok(Tx::with_txid(TxId::from_tx(&tx), tx))
    }
}

impl BitcoinSer for OutPoint {
    fn ser_to<S: BitcoinSerializer>(&self, bytes: &mut S) {
        self.txid.ser_to(bytes);
        self.out_idx.ser_to(bytes);
    }

    fn deser(data: &mut bytes::Bytes) -> Result<Self, DataError> {
        Ok(OutPoint {
            txid: BitcoinSer::deser(data)?,
            out_idx: BitcoinSer::deser(data)?,
        })
    }
}

impl BitcoinSer for TxInput {
    fn ser_to<S: BitcoinSerializer>(&self, bytes: &mut S) {
        self.prev_out.ser_to(bytes);
        self.script.ser_to(bytes);
        self.sequence.ser_to(bytes);
    }

    fn deser(data: &mut bytes::Bytes) -> Result<Self, DataError> {
        Ok(TxInput {
            prev_out: BitcoinSer::deser(data)?,
            script: BitcoinSer::deser(data)?,
            sequence: BitcoinSer::deser(data)?,
            coin: None,
        })
    }
}

impl BitcoinSer for TxOutput {
    fn ser_to<S: BitcoinSerializer>(&self, bytes: &mut S) {
        self.value.ser_to(bytes);
        self.script.ser_to(bytes);
    }

    fn deser(data: &mut bytes::Bytes) -> Result<Self, DataError> {
        Ok(TxOutput {
            value: BitcoinSer::deser(data)?,
            script: BitcoinSer::deser(data)?,
        })
    }
}

#[cfg(test)]
mod tests {
    use bytes::Bytes;

    use crate::{
        script::Script,
        ser::BitcoinSer,
        tx::{OutPoint, Tx, TxId, TxInput, TxMut, TxOutput},
    };

    fn verify_ser(tx: TxMut, ser: &[u8]) {
        assert_eq!(tx.ser().as_ref(), ser);
        assert_eq!(tx.ser_len(), ser.len());
        let mut bytes = Bytes::copy_from_slice(ser);
        assert_eq!(tx, TxMut::deser(&mut bytes).unwrap());

        let tx = Tx::with_txid(TxId::from_tx(&tx), tx);
        assert_eq!(tx.ser().as_ref(), ser);
        assert_eq!(tx.ser_len(), ser.len());
        let mut bytes = Bytes::copy_from_slice(ser);
        assert_eq!(tx, Tx::deser(&mut bytes).unwrap());
    }

    #[test]
    fn test_ser_tx() -> Result<(), hex::FromHexError> {
        verify_ser(TxMut::default(), &[0; 10]);
        verify_ser(
            TxMut {
                version: 0x12345678,
                inputs: vec![],
                outputs: vec![],
                locktime: 0x9abcdef1,
            },
            &hex::decode("785634120000f1debc9a")?,
        );
        let genesis_tx = TxMut {
            version: 1,
            inputs: vec![TxInput {
                prev_out: OutPoint {
                    txid: TxId::from([0; 32]),
                    out_idx: 0xffff_ffff,
                },
                script: Script::new(
                    hex::decode(
                        "04ffff001d0104455468652054696d65732030332f4a616e2f3230\
                         3039204368616e63656c6c6f72206f6e206272696e6b206f662073\
                         65636f6e64206261696c6f757420666f722062616e6b73",
                    )?
                    .into(),
                ),
                sequence: 0xffff_ffff,
                coin: None,
            }],
            outputs: vec![TxOutput {
                value: 5000000000,
                script: Script::new(
                    hex::decode(
                        "4104678afdb0fe5548271967f1a67130b7105cd6a828e03909a679\
                         62e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384df7\
                         ba0b8d578a4c702b6bf11d5fac",
                    )?
                    .into(),
                ),
            }],
            locktime: 0,
        };
        verify_ser(
            genesis_tx,
            &hex::decode(
                "01000000010000000000000000000000000000000000000000000000000000\
                 000000000000ffffffff4d04ffff001d0104455468652054696d6573203033\
                 2f4a616e2f32303039204368616e63656c6c6f72206f6e206272696e6b206f\
                 66207365636f6e64206261696c6f757420666f722062616e6b73ffffffff01\
                 00f2052a01000000434104678afdb0fe5548271967f1a67130b7105cd6a828\
                 e03909a67962e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384d\
                 f7ba0b8d578a4c702b6bf11d5fac00000000"
            )?,
        );
        Ok(())
    }
}

File: modules/ecash-lib/src/tx.ts
// Copyright (c) 2024 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

import { Bytes } from './io/bytes.js';
import { fromHexRev, toHexRev, fromHex, toHex } from './io/hex.js';
import { writeVarSize, readVarSize } from './io/varsize.js';
import { Writer } from './io/writer.js';
import { WriterBytes } from './io/writerbytes.js';
import { WriterLength } from './io/writerlength.js';
import { Ecc } from './ecc.js';
import { Script } from './script.js';
import { sha256d } from './hash.js';
import { flagSignature } from './signatories.js';
import { ALL_BIP143, SigHashType } from './sigHashType.js';
import { UnsignedTx } from './unsignedTx.js';

/**
 * Default value for nSequence of inputs if left undefined; this opts out of
 * BIP68 relative lock-time, and if all inputs have this value, nLockTime is
 * disabled, too.
 *
 * This is chosen as the default as it's the default in the node too,
 * see CTxIn in /src/primitives/transaction.h.
 **/
export const DEFAULT_SEQUENCE = 0xffffffff;

/** Current tx version, see CTransaction in /stc/primitives/transaction.h */
export const DEFAULT_TX_VERSION = 2;

/** COutPoint, pointing to a coin being spent. */
export interface OutPoint {
    /**
     * TxId of the output of the coin, either a big-endian hex encoded TxId
     * or a little-endian bytearray.
     **/
    txid: string | Uint8Array;
    /** Index in the outputs of the tx of the coin. */
    outIdx: number;
}

/** CTxIn, spending an unspent output. */
export interface TxInput {
    /** Points to an output being spent. */
    prevOut: OutPoint;
    /** scriptSig unlocking the output, defaults to the empty Script. */
    script?: Script;
    /** nSequence, defaults to 0xffffffff if unspecified. */
    sequence?: number;
    /** Sign data required to sign an input */
    signData?: SignData;
}

/** CTxOut, creating a new output. */
export interface TxOutput {
    /** Value in satoshis of the output (1 XEC = 100 satoshis) */
    sats: bigint;
    /** Script locking the output */
    script: Script;
}

/** All the data required to sign an input (using BIP143). */
export interface SignData {
    /** Value of the output being spent */
    sats: bigint;
    /** Script of the output being spent (not for P2SH) */
    outputScript?: Script;
    /**
     * For P2SH inputs, the preimage of the script hash locking the output being
     * spent.
     **/
    redeemScript?: Script;
}

/** CTransaction, a Bitcoin transaction. */
export class Tx {
    /** nVersion of the tx */
    public version: number;
    /** vin, tx inputs spending outputs of other txs */
    public inputs: TxInput[];
    /** vout, tx outputs created in this tx */
    public outputs: TxOutput[];
    /** nLockTime of the tx, specifies when the tx can be mined earliest */
    public locktime: number;

    public constructor(params?: {
        version?: number;
        inputs?: TxInput[];
        outputs?: TxOutput[];
        locktime?: number;
    }) {
        this.version = params?.version ?? DEFAULT_TX_VERSION;
        this.inputs = params?.inputs ?? [];
        this.outputs = params?.outputs ?? [];
        this.locktime = params?.locktime ?? 0;
    }

    /** Serialize the tx to a byte array */
    public ser(): Uint8Array {
        const writerBytes = new WriterBytes(this.serSize());
        this.write(writerBytes);
        return writerBytes.data;
    }

    /** Serialize the tx to a hex string */
    public toHex(): string {
        return toHex(this.ser());
    }

    /** Calculate the serialized size of the tx */
    public serSize(): number {
        const writerLength = new WriterLength();
        this.write(writerLength);
        return writerLength.length;
    }

    /** Write the tx to the given writer */
    public write(writer: Writer) {
        writer.putU32(this.version);
        writeVarSize(this.inputs.length, writer);
        for (const input of this.inputs) {
            writeTxInput(input, writer);
        }
        writeVarSize(this.outputs.length, writer);
        for (const output of this.outputs) {
            writeTxOutput(output, writer);
        }
        writer.putU32(this.locktime);
    }

    /** Deserialize a Tx from a Uint8Array */
    public static deser(data: Uint8Array): Tx {
        const bytes = new Bytes(data);
        const version = bytes.readU32();
        const numInputs = readVarSize(bytes);
        const inputs: TxInput[] = [];
        for (let i = 0; i < numInputs; ++i) {
            // Read OutPoint
            const txid = bytes.readBytes(32);
            const outIdx = bytes.readU32();
            // Read script
            const script = Script.readWithSize(bytes);
            // Read sequence
            const sequence = bytes.readU32();
            inputs.push({
                prevOut: {
                    txid,
                    outIdx,
                },
                script,
                sequence,
            });
        }
        const numOutputs = readVarSize(bytes);
        const outputs: TxOutput[] = [];
        for (let i = 0; i < numOutputs; ++i) {
            outputs.push(readTxOutput(bytes));
        }
        const locktime = bytes.readU32();
        return new Tx({
            version,
            inputs,
            outputs,
            locktime,
        });
    }

    /** Deserialize a Tx from a hex string */
    public static fromHex(hex: string): Tx {
        return Tx.deser(fromHex(hex));
    }

    /**
     * Compute the transaction ID (TxId) as a hex string (little-endian).
     * This follows the eCash convention: the TxId is the double SHA256 of the
     * serialized transaction, returned as a hex string in little-endian (reversed) order.
     * See the node src/primitives/txid.h for more details.
     */
    public txid(): string {
        return toHexRev(sha256d(this.ser()));
    }

    /**
     * Attempt to parse a **non-SegWit** serialized transaction from `data` — the same
     * encoding as {@link Tx.deser} / `Tx.ser()` (version, inputs, outputs, locktime;
     * no witness marker or witness stacks). eCash does not use SegWit transactions,
     * so this is the full-transaction wire format on-chain here. Returns a `Tx` only
     * when `data` is **exactly** one such transaction: the parse must consume the
     * **entire** buffer (no trailing bytes). Returns `undefined` on malformed input
     * or if any bytes remain after `locktime`.
     *
     * **PSBT-only motivation:** Bitcoin ABC’s PSBT input key `0x00` (`PSBT_IN_UTXO`)
     * stores **either** a full previous transaction (BIP 174 “non-witness UTXO”) **or**
     * a compact `CTxOut` (amount + `scriptPubKey`). Callers must disambiguate. Plain
     * {@link Tx.deser} reads a tx from the start of `data` but **does not** require
     * `data.length` to match the serialized length — leftover bytes are ignored, so
     * you cannot use it to prove “this blob is solely a full tx.” This helper is
     * used from PSBT (`resolveWitnessFromKey00` in `psbt.ts`): if `tryDeserExact`
     * succeeds (and the txid matches), treat as non-witness UTXO; otherwise decode as
     * `CTxOut`-shaped bytes.
     */
    public static tryDeserExact(data: Uint8Array): Tx | undefined {
        try {
            const bytes = new Bytes(data);
            const version = bytes.readU32();
            const numInputs = readVarSize(bytes);
            const inputs: TxInput[] = [];
            for (let i = 0; i < numInputs; ++i) {
                const txid = bytes.readBytes(32);
                const outIdx = bytes.readU32();
                const script = Script.readWithSize(bytes);
                const sequence = bytes.readU32();
                inputs.push({
                    prevOut: { txid, outIdx },
                    script,
                    sequence,
                });
            }
            const numOutputs = readVarSize(bytes);
            const outputs: TxOutput[] = [];
            for (let i = 0; i < numOutputs; ++i) {
                outputs.push(readTxOutput(bytes));
            }
            const locktime = bytes.readU32();
            if (bytes.idx !== data.length) {
                return undefined;
            }
            return new Tx({ version, inputs, outputs, locktime });
        } catch {
            return undefined;
        }
    }

    /**
     * Add a signature to a partially-signed multisig input.
     * Verifies the signature against the sighash for each pubkey in the
     * redeem/output script and merges with existing signatures (which pubkey
     * signed is inferred from verification).
     */
    public addMultisigSignature(params: {
        inputIdx: number;
        signature: Uint8Array;
        signData: SignData;
        ecc?: Ecc;
    }): Tx {
        const { inputIdx, signature, signData } = params;
        const ecc = params.ecc ?? new Ecc();
        const input = this.inputs[inputIdx];
        if (!input.script || input.script.bytecode.length === 0) {
            throw new Error(
                `Input ${inputIdx} has no scriptSig to add signature to`,
            );
        }
        const isBare =
            signData.outputScript !== undefined &&
            signData.redeemScript === undefined;
        const parsed = isBare
            ? input.script.parseBareMultisigSpend(signData.outputScript!)
            : input.script.parseP2shMultisigSpend();
        const txWithSignData = new Tx({
            version: this.version,
            inputs: this.inputs.map((inp, i) =>
                i === inputIdx
                    ? { ...copyTxInput(inp), signData }
                    : copyTxInput(inp),
            ),
            outputs: this.outputs,
            locktime: this.locktime,
        });
        const unsignedTx = UnsignedTx.fromTx(txWithSignData);
        const inputAt = unsignedTx.inputAt(inputIdx);
        const sigHashType =
            SigHashType.fromInt(
                (signature[signature.length - 1] ?? 0) & 0xff,
            ) ?? ALL_BIP143;
        const preimage = inputAt.sigHashPreimage(sigHashType);
        const sighash = sha256d(preimage.bytes);
        const sigWithoutFlag = signature.slice(0, -1);

        let pubkeyIndex = -1;
        if (parsed.isSchnorr) {
            for (let i = 0; i < parsed.pubkeys.length; i++) {
                try {
                    ecc.schnorrVerify(
                        sigWithoutFlag,
                        sighash,
                        parsed.pubkeys[i]!,
                    );
                    pubkeyIndex = i;
                    break;
                } catch {
                    /* try next pubkey */
                }
            }
            if (pubkeyIndex < 0) {
                throw new Error(
                    'Schnorr signature does not verify for any pubkey in the multisig script',
                );
            }
        } else {
            for (let i = 0; i < parsed.pubkeys.length; i++) {
                try {
                    ecc.ecdsaVerify(
                        sigWithoutFlag,
                        sighash,
                        parsed.pubkeys[i]!,
                    );
                    pubkeyIndex = i;
                    break;
                } catch {
                    /* try next pubkey */
                }
            }
            if (pubkeyIndex < 0) {
                throw new Error(
                    'ECDSA signature does not verify for any pubkey in the multisig script',
                );
            }
        }

        const sigsByPubkey: (Uint8Array | undefined)[] = Array(
            parsed.pubkeys.length,
        ).fill(undefined);

        if (parsed.isSchnorr) {
            const indices = parsed.pubkeyIndices!;
            const sortedIndices = [...indices].sort((a, b) => a - b);
            for (let i = 0; i < parsed.signatures.length; i++) {
                const sig = parsed.signatures[i];
                if (sig !== undefined && i < sortedIndices.length) {
                    sigsByPubkey[sortedIndices[i]!] = sig;
                }
            }
        } else {
            for (const sig of parsed.signatures) {
                if (sig === undefined) continue;
                const sigNoFlag = sig.slice(0, -1);
                for (let i = 0; i < parsed.pubkeys.length; i++) {
                    try {
                        ecc.ecdsaVerify(sigNoFlag, sighash, parsed.pubkeys[i]!);
                        sigsByPubkey[i] = sig;
                        break;
                    } catch {
                        /* try next pubkey */
                    }
                }
            }
        }
        sigsByPubkey[pubkeyIndex] = signature;

        const nonNullSigs = sigsByPubkey.filter(
            (s): s is Uint8Array => s !== undefined,
        );
        const sigsForScript =
            nonNullSigs.length >= parsed.numSignatures
                ? nonNullSigs.slice(0, parsed.numSignatures)
                : [
                      ...nonNullSigs,
                      ...Array(parsed.numSignatures - nonNullSigs.length).fill(
                          undefined,
                      ),
                  ];

        const redeemScript: Script | undefined =
            !isBare && 'redeemScript' in parsed
                ? (parsed as { redeemScript: Script }).redeemScript
                : undefined;
        const newScriptSig = parsed.isSchnorr
            ? (() => {
                  const signerIndices = new Set<number>();
                  for (
                      let i = 0;
                      i < parsed.pubkeys.length &&
                      signerIndices.size < parsed.numSignatures;
                      i++
                  ) {
                      if (sigsByPubkey[i] !== undefined) signerIndices.add(i);
                  }
                  return isBare
                      ? Script.multisigSpend({
                            signatures: sigsForScript,
                            pubkeyIndices: signerIndices,
                            numPubkeys: parsed.numPubkeys,
                        })
                      : Script.multisigSpend({
                            signatures: sigsForScript,
                            redeemScript,
                            pubkeyIndices: signerIndices,
                        });
              })()
            : Script.multisigSpend({
                  signatures: sigsForScript,
                  redeemScript,
              });

        const newInputs = this.inputs.map((inp, i) =>
            i === inputIdx
                ? { ...copyTxInput(inp), script: newScriptSig }
                : copyTxInput(inp),
        );
        return new Tx({
            version: this.version,
            inputs: newInputs,
            outputs: this.outputs,
            locktime: this.locktime,
        });
    }

    /**
     * Like {@link addMultisigSignature}, but computes the signature from a
     * secret key: BIP143 preimage (or legacy if `sigHashType` is legacy),
     * Schnorr for Schnorr-format multisig spends and ECDSA otherwise.
     */
    public addMultisigSignatureFromKey(params: {
        inputIdx: number;
        sk: Uint8Array;
        signData: SignData;
        /** Defaults to {@link ALL_BIP143}. */
        sigHashType?: SigHashType;
        ecc?: Ecc;
    }): Tx {
        const sigHashType = params.sigHashType ?? ALL_BIP143;
        const ecc = params.ecc ?? new Ecc();
        const { inputIdx, sk, signData } = params;
        const input = this.inputs[inputIdx];
        if (!input.script || input.script.bytecode.length === 0) {
            throw new Error(
                `Input ${inputIdx} has no scriptSig to add signature to`,
            );
        }
        const isBare =
            signData.outputScript !== undefined &&
            signData.redeemScript === undefined;
        const parsed = isBare
            ? input.script.parseBareMultisigSpend(signData.outputScript!)
            : input.script.parseP2shMultisigSpend();
        const txWithSignData = new Tx({
            version: this.version,
            inputs: this.inputs.map((inp, i) =>
                i === inputIdx
                    ? { ...copyTxInput(inp), signData }
                    : copyTxInput(inp),
            ),
            outputs: this.outputs,
            locktime: this.locktime,
        });
        const unsignedTx = UnsignedTx.fromTx(txWithSignData);
        const preimage = unsignedTx
            .inputAt(inputIdx)
            .sigHashPreimage(sigHashType);
        const sighash = sha256d(preimage.bytes);
        const sig = parsed.isSchnorr
            ? ecc.schnorrSign(sk, sighash)
            : ecc.ecdsaSign(sk, sighash);
        const signature = flagSignature(sig, sigHashType);
        return this.addMultisigSignature({
            inputIdx,
            signature,
            signData,
            ecc,
        });
    }

    /**
     * Whether every **multisig** input (identified from `signData`) has enough
     * signatures in its scriptSig. Non-multisig inputs are ignored.
     *
     * If the transaction has **no** multisig inputs, this returns `true` (there
     * is nothing multisig-specific left to satisfy). That can look surprising on
     * a non-multisig or otherwise incomplete tx; this helper is **not** a
     * broadcast-readiness check. Call sites are expected to use it only in
     * multisig / PSBT flows where the question is specifically whether multisig
     * inputs still need more signatures (including mixed txs: non-multisig
     * inputs are finalized elsewhere).
     */
    public isFullySignedMultisig(): boolean {
        for (let i = 0; i < this.inputs.length; i++) {
            const input = this.inputs[i];
            const multisigScript =
                input.signData?.redeemScript !== undefined
                    ? input.signData.redeemScript
                    : input.signData?.outputScript?.isMultisig()
                      ? input.signData!.outputScript
                      : undefined;
            if (multisigScript === undefined) {
                continue;
            }
            if (!input.script || input.script.bytecode.length === 0) {
                return false;
            }
            try {
                const parsed =
                    input.signData?.redeemScript === undefined
                        ? input.script.parseBareMultisigSpend(multisigScript)
                        : input.script.parseP2shMultisigSpend();
                const sigCount = parsed.signatures.filter(
                    s => s !== undefined,
                ).length;
                if (sigCount < parsed.numSignatures) {
                    return false;
                }
            } catch {
                return false;
            }
        }
        return true;
    }
}

export function readTxOutput(bytes: Bytes): TxOutput {
    const sats = bytes.readU64();
    const script = Script.readWithSize(bytes);
    return {
        sats,
        script,
    };
}

/** Write an outpoint to a Writer */
export function writeOutPoint(outpoint: OutPoint, writer: Writer): void {
    const txid =
        typeof outpoint.txid === 'string'
            ? fromHexRev(outpoint.txid)
            : outpoint.txid;
    writer.putBytes(txid);
    writer.putU32(outpoint.outIdx);
}

/** Write a TxInput to a Writer */
export function writeTxInput(input: TxInput, writer: Writer): void {
    writeOutPoint(input.prevOut, writer);
    (input.script ?? new Script()).writeWithSize(writer);
    writer.putU32(input.sequence ?? DEFAULT_SEQUENCE);
}

/** Write a TxOutput to a Writer */
export function writeTxOutput(output: TxOutput, writer: Writer): void {
    writer.putU64(output.sats);
    output.script.writeWithSize(writer);
}

/** Create a deep copy of the TxInput */
export function copyTxInput(input: TxInput): TxInput {
    return {
        prevOut: {
            txid:
                typeof input.prevOut.txid === 'string'
                    ? input.prevOut.txid
                    : new Uint8Array(input.prevOut.txid),
            outIdx: input.prevOut.outIdx,
        },
        script: input.script?.copy(),
        sequence: input.sequence,
        signData: input.signData && {
            sats: input.signData.sats,
            outputScript: input.signData.outputScript?.copy(),
            redeemScript: input.signData.redeemScript?.copy(),
        },
    };
}

/** Create a deep copy of the TxOutput */
export function copyTxOutput(output: TxOutput): TxOutput {
    return {
        sats: output.sats,
        script: output.script.copy(),
    };
}
